ledger/db/src/mint_config_store.rs

```diff
@@ -0,0 +1,573 @@
+//! configuration. This is used to enforce the per-configuration mint limit. 2)
+//! A mapping of nonce -> SetMintConfigTx object containing the nonce. This is
+//! mainly used to prevent replay attacks.
+
+use crate::{u32_to_key_bytes, Error};
+use lmdb::{Database, DatabaseFlags, Environment, RwTransaction, Transaction, WriteFlags};
+use mc_transaction_core::{
+    mint::{MintConfig, SetMintConfigTx},
+    TokenId,
+};
+use mc_util_serial::{decode, encode, Message};
+
+// LMDB Database names.
+pub const ACTIVE_MINT_CONFIGS_BY_TOKEN_ID_DB_NAME: &str =
+    "mint_config_store:active_mint_configs_by_token_id";
+pub const SET_MINT_CONFIG_TX_BY_NONCE_DB_NAME: &str =
+    "mint_config_store:set_mint_config_tx_by_nonce_db_name";
+
+/// An active mint configuration for a single token.
+#[derive(Clone, Eq, Message, PartialEq)]
+pub struct ActiveMintConfig {
+    /// The actual mint configuration.
+    #[prost(message, required, tag = "1")]
+    pub mint_config: MintConfig,
+
+    /// How many tokens have been minted using this configuration.
+    #[prost(uint64, tag = "2")]
+    pub total_minted: u64,
+}
+
+/// A collection of active mint configurations.
+/// This is needed for serializing/deserializing a Vec<ActiveMintConfig>.
+#[derive(Clone, Eq, Message, PartialEq)]
+struct ActiveMintConfigs {
+    #[prost(message, repeated, tag = "1")]
+    pub configs: Vec<ActiveMintConfig>,
+}
+
+impl From<&SetMintConfigTx> for ActiveMintConfigs {
+    fn from(set_mint_config_tx: &SetMintConfigTx) -> Self {
+        ActiveMintConfigs {
+            configs: set_mint_config_tx
+                .prefix
+                .configs
+                .iter()
+                .map(|mint_config| ActiveMintConfig {
+                    mint_config: mint_config.clone(),
+                    total_minted: 0,
+                })
+                .collect(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct MintConfigStore {
+    /// token id -> Vec<ActiveMintConfig>
+    active_mint_configs_by_token_id: Database,
+
+    /// nonce -> SetMintConfigTx
+    set_mint_config_tx_by_nonce: Database,
+}
+
+impl MintConfigStore {
+    /// Opens an existing MintConfigStore.
+    pub fn new(env: &Environment) -> Result<Self, Error> {
+        Ok(MintConfigStore {
+            active_mint_configs_by_token_id: env
+                .open_db(Some(ACTIVE_MINT_CONFIGS_BY_TOKEN_ID_DB_NAME))?,
+            set_mint_config_tx_by_nonce: env.open_db(Some(SET_MINT_CONFIG_TX_BY_NONCE_DB_NAME))?,
+        })
+    }
+
+    /// Creates a fresh MintConfigStore.
+    pub fn create(env: &Environment) -> Result<(), Error> {
+        env.create_db(
+            Some(ACTIVE_MINT_CONFIGS_BY_TOKEN_ID_DB_NAME),
+            DatabaseFlags::empty(),
+        )?;
+        env.create_db(
+            Some(SET_MINT_CONFIG_TX_BY_NONCE_DB_NAME),
+            DatabaseFlags::empty(),
+        )?;
+        Ok(())
+    }
+
+    /// Set mint configurations for a given token.
+    pub fn set_active_mint_configs(
```

Comment:
Sorry I was a little confused by the intention of this. Could we add a test that demonstrates the expected behaviour when we deactivate a mint config via this method or reactivate a mint config via this method?

Reply:
I believe this is covered by `set_get_behaves_correctly`

Reply:
I believe that set_get_behaves_correctly is testing setting the mint config and retrieving that same config for 2 different tokens. What I'm unsure about is the behaviour of deactivating an active config on a token, e.g. we go from having an active mint config to an empty active mint config, perhaps after having already minted some tokens. Then reactivating the previously deactivated config.

Reply:
Minting configurations can be replaced at any time and do not affect anything previously minted, they just control who can currently mint. There is no deactivation - other than by replacing the currently active configuration with an empty one`set_replaces_configuration` checks that an existing configuration can be replaced by a new one. I'll add a test that verifies an empty one is accepted.